### ISSUE TITLE
Constant configmap updates causing nonstop handler runs

### ISSUE THREAD

**Issue by rmweir** [rancher/rancher#36528]
**Rancher Server Setup**
- Rancher version:
2.6.3+

**Describe the bug**
Configmap lasso handler are running constantly due to certain configmaps being edited frequently. Observed one configmap being updated every 2 seconds. These tasks should be fullfilled:
* Investigate why these confimaps are constantly being updated.
* Is it a possibility for these configmaps to not be updated on a fixed interval?
* Investigate what these configmap handlers do and if there is an alternative to having handlers watching configmap events.
* Deregister configmap manually and see if this has an effect on cpu/memory footprint.

**To Reproduce**
 
Install rancher, then install monitoring using steps https://github.com/rancher/rancher/wiki/How-to-setup-prometheus-dashboard-for-controllers. Metrics for lasso handlers will be available soon once https://github.com/rancher/lasso/pull/17 is merged and brought into rancher.

**Result**
Config map handlers are never reaching a plateu and are running constantly:
![image](https://user-images.githubusercontent.com/19376037/154164051-675d4917-3ef9-4e24-bb2c-fc2fbc6e39e3.png)

I watched the configmaps and saw some were having constant events. Here are a few events from one:
```
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"gitjob-6cd5874f8c-q8zz7","leaseDurationSeconds":45,"acquireTime":"2022-02-04T22:35:42Z","renewTime":"2022-02-08T02:06:32Z","leaderTransitions":1}'
  creationTimestamp: "2022-02-04T18:20:20Z"
  name: gitjob
  namespace: cattle-fleet-system
  resourceVersion: "1106478"
  uid: 8d59a318-86d5-4567-964d-6f9c9a55e5c2
---
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"gitjob-6cd5874f8c-q8zz7","leaseDurationSeconds":45,"acquireTime":"2022-02-04T22:35:42Z","renewTime":"2022-02-08T02:06:34Z","leaderTransitions":1}'
  creationTimestamp: "2022-02-04T18:20:20Z"
  name: gitjob
  namespace: cattle-fleet-system
  resourceVersion: "1106486"
  uid: 8d59a318-86d5-4567-964d-6f9c9a55e5c2
---
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"gitjob-6cd5874f8c-q8zz7","leaseDurationSeconds":45,"acquireTime":"2022-02-04T22:35:42Z","renewTime":"2022-02-08T02:06:36Z","leaderTransitions":1}'
  creationTimestamp: "2022-02-04T18:20:20Z"
  name: gitjob
  namespace: cattle-fleet-system
  resourceVersion: "1106493"
  uid: 8d59a318-86d5-4567-964d-6f9c9a55e5c2
---
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"gitjob-6cd5874f8c-q8zz7","leaseDurationSeconds":45,"acquireTime":"2022-02-04T22:35:42Z","renewTime":"2022-02-08T02:06:38Z","leaderTransitions":1}'
  creationTimestamp: "2022-02-04T18:20:20Z"
  name: gitjob
  namespace: cattle-fleet-system
  resourceVersion: "1106501"
  uid: 8d59a318-86d5-4567-964d-6f9c9a55e5c2
```
It appears that the leader annotations is being updated to to the renewTime field.

**Expected Result**

These handlers should reach a plateau or if these change events are non negotiable, we may want to find alternatives to using configmap handlers.

**Comment by samjustus** [comment]
try disabling fleet to see if that makes this issue go away

**Comment by cmurphy** [comment]
The top handlers that are firing off at an upward slope are `snapshotbackpopulate`, `helm-app-configmap` and `prov-log-configmap` which are all configmap handlers. The configmaps that are being updated so frequently are `kube-system/cattle-controllers`, `cattle-fleet-system/gitjob`, `cattle-fleet-system/fleet-controller-lock`, and `cattle-fleet-local-system/fleet-agent-lock`. These are all leader election configmaps and the updates are all for renewing leases. The `RetryPeriod` for the leader election config is [always 2 seconds](https://github.com/rancher/wrangler/blob/a4503aad84eb1021c0b1c3a24c30df790fd68584/pkg/leader/leader.go#L55) (except in dev mode) plus a jitter factor, and this value is used as the [polling period](https://github.com/rancher/client-go/blob/e4f74dbbebfe753d2b9a45b6d1c661697548f0e8/tools/leaderelection/leaderelection.go#L271) in the leadership renewal. This is working as expected, and configmap handlers firing off frequently is a natural consequence. We could change the renewal period to a higher value or make it configurable to try to reduce the slope, but the increase over time is unavoidable.

operator-sdk offers an [alternate type of leader election](https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#leader-election) that uses owner references rather than annotations so that the leader config object is garbage collected when the leader dies rather than updated at every renewal period. It comes with its own tradeoffs, and is not part of the upstream leader election implementation, but using that framework or using its implementation as inspiration is an option.

I commented out the registration of those three handlers and saw no discernible difference in the CPU or memory utilization with these controllers turned off (the spike is the rancher restart):

![image](https://user-images.githubusercontent.com/1220710/166323183-0dc13a88-6954-467d-bf06-a7f67b619644.png)

IMO we don't need to make changes to this until we can show that it is causing a problem.
